compiler/parser/RdrHsSyn.hs

```diff
@@ -2381,19 +2415,25 @@ mkTemplateClassInstanceSigs templateName mbKeyType =
     mkSig methodName ty =
       let fullMethodName = noLoc $ mkRdrUnqual $ mkVarOcc $ prefixTemplateClassMethod $ methodName ++ templateName
       in  noLoc $ ClassOpSig noExt False [fullMethodName] (HsIB noExt ty)
-    templateType = mkUnqualType templateName
-    contractId = mkContractId templateType
-    hasKeyType = mkQualType "HasKey" `mkAppTy` templateType
+    templateType = mkTemplateType templateName tyVars
```

Comment:
Instead of always applying `mbParenTy` to `templateType`, we should do it here:
```suggestion
    templateType = (if null tyVars then id else mkParenTy) $ mkTemplateType templateName tyVars
```

Reply:
Actually, this happens in other places as well. Should be add the parenthesizing into `mkTemplateType`?

Reply:
I was trying to avoid unnecessary parentheses, which you can only determine at the occurrences. I could add add parens in `mkTemplateType`, but you'd end up with lots of extra (harmless) parens.

Reply:
We could add a flag to `templateType` which determines whether we want to add parentheses when they are necessary.